File: CS/Basic/sort/margeSort.cpp
#include <iostream>
#include <string>
#include <cmath>
#include <vector>
#include <iterator>
#include <sstream>
#include <climits>

using namespace std;

vector<int> mergeSort(vector<int> arr)
{
    if (arr.size() > 1)
    {
        int middle = arr.size() / 2;

        vector<int> leftArr(arr.begin(), arr.begin() + middle);
        vector<int> rightArr(arr.begin() + middle, arr.begin() + arr.size());

        leftArr = mergeSort(leftArr);
        rightArr = mergeSort(rightArr);

        int leftIndex = 0;
        int rightIndex = 0;
        int insertIndex = 0;

        leftArr.push_back(INT_MAX);
        rightArr.push_back(INT_MAX);

        while (leftIndex + rightIndex < arr.size())
        {
            if (leftArr[leftIndex] <= rightArr[rightIndex])
            {
                arr[insertIndex] = leftArr[leftIndex];
                leftIndex++;
            }
            else
            {
                arr[insertIndex] = rightArr[rightIndex];
                rightIndex++;
            }
            insertIndex++;
        }
    }

    return arr;
}

File: CS/Basic/sort/margeSort.py
def mergeSort(arr: list[int]) -> list[int]:
    if len(arr) <= 1:
        return arr
    middle = len(arr) // 2
    leftArr = mergeSort(arr[0:middle])
    rightArr = mergeSort(arr[middle:])

    leftIndex, rightIndex, key = 0, 0, 0

    while leftIndex < len(leftArr) and rightIndex < len(rightArr):
        if leftArr[leftIndex] <= rightArr[rightIndex]:
            arr[key] = leftArr[leftIndex]
            leftIndex += 1
        else:
            arr[key] = rightArr[rightIndex]
            rightIndex += 1

        key += 1

    while leftIndex < len(leftArr):
        arr[key] = leftArr[leftIndex]
        leftIndex += 1
        key += 1

    while rightIndex < len(rightArr):
        arr[key] = rightArr[rightIndex]
        rightIndex += 1
        key += 1

    return arr


print(mergeSort([2, 12, 5, 10, 9, 8]))
